acceptancetests/assess_primary_sub_relations.py

```diff
@@ -0,0 +1,225 @@
+
+    # We expect this check to fail now as the subordinate app shouldn't be
+    # running.
+    # If it doesn't fail then the unit is still operational and that is a fail
+    # state.
+    try:
+        ensure_subordinate_unit_operational(
+            test_model, primary_app_name='dummy-source')
+    except JujuAssertionError:
+        log.info('Subordinate application relation successfully removed.')
+    else:
+        raise JujuAssertionError(
+            'Subordinate unit still operational after relation removal')
+
+    test_model.destroy_model()
+
+
+def assert_service_is_removed(client, application_name, timeout=60):
+    """Wait for the named application to be removed from status.
+
+    Check status of `client` until the named application `application_name` no
+    longer appears.
+    Will raise if the application still appears after the timeout.abs
+
+    :param client: ModelClient with which to check for application status.
+    :param application_name: String name of the application to assert for in
+      status.
+    :param timeout: int seconds to wait for raising exception if application
+      still exists.
+    :raises JujuAssertionError: When application_name still appears in status
+      after the timeout is encountered.
+    """
+
+    for _ in until_timeout(timeout):
+        current_status = client.get_status()
+        current_status.get_applications().keys()
+        if application_name not in current_status.get_applications().keys():
+            return
+        else:
+            time.sleep(1)
+    raise JujuAssertionError(
+        'Application "{}" failed to be removed in {} seconds'.format(
+            application_name, timeout))
+
+
+def assert_subordinate_is_removed(
+        client, primary_name, subordinate_name, timeout=60):
+    for _ in until_timeout(timeout):
+        status = client.get_status()
+        subordinate_names = [
+            unit[0] for unit in status.service_subordinate_units(
+                primary_name)
+            ]
+        if not any(subordinate_name in unit for unit in subordinate_names):
+            return
+        else:
+            time.sleep(1)
+    raise JujuAssertionError(
+        'Subordinate "{}" failed to be removed from "{}" in {} seconds'.format(
+            subordinate_name, primary_name, timeout))
+
+
+def create_model_with_primary_and_sub_apps(client, model_name):
+    test_model = client.add_model(model_name)
+    deploy_primary_charm(test_model, 'dummy-source')
+    deploy_primary_charm(test_model, 'dummy-sink')
+
+    test_model.deploy('cs:ntp')
+
+    test_model.set_config('dummy-source', {'token': 'subordinate-test'})
+    test_model.juju('add-relation', ('dummy-source', 'dummy-sink'))
+    test_model.juju('add-relation', ('ntp', 'dummy-source'))
+    test_model.juju('add-relation', ('ntp', 'dummy-sink'))
+    test_model.wait_for_workloads()
+    return test_model
+
+
+def ensure_subordinate_unit_operational(client, primary_app_name):
```

Comment:
You've used ensure and assert language -- what's the difference? This ensure method still raises an assertion error, so I'm not sure what difference is intended in the ensure* methods vs assert* methods.